src/main/java/com/exactpro/th2/taskutils/FutureTracker.java

```diff
@@ -23,71 +27,181 @@
-            synchronized (futures) {
-                futures.add(future);
-            }
-            future.whenComplete((res, ex) -> {
-                synchronized (futures) {
-                    futures.remove(future);
+            long endNanos = System.nanoTime() + timeout.toNanos();
+
+            lock.writeLock().lock();
+            try {
+                if (future.isDone()) {
+                    return true;
+                }
+                if (!hasLimit || futures.size() < limit) {
+                    addFuture(future);
+                    return true;
+                }
+
+                long remainingTime = endNanos - System.nanoTime();
+                if (remainingTime <= 0) {
+                    return future.isDone();
+                }
+
+                if (!onRemove.await(remainingTime, TimeUnit.NANOSECONDS)) {
+                    return future.isDone();
+                }
+                if (future.isDone()) {
+                    return true;
                 }
-            });
+                if (futures.size() < limit) {
+                    addFuture(future);
+                    return true;
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
         }
+        throw new IllegalStateException("Future tracker is already disabled");
     }
 
     /**
      * Stops accepting new futures and waits for
      * currently tracked futures
      */
-    public void awaitRemaining () {
-        awaitRemaining(-1);
+    public void awaitRemaining() throws InterruptedException {
+        awaitRemaining(Duration.ofMillis(-1));
     }
 
     /**
-     * Waits for tracked futures, cancels futures after timeout.
+     * Disables traker and waits for tracked futures, cancels futures after timeout.
      * Cancels all futures without wait if passed 0.
      * Waits without timeout if passed negative.
+     *
      * @param timeoutMillis milliseconds to wait
+     * @deprecated use {@link #awaitRemaining(Duration)} instead
+     */
+    @Deprecated
+    public void awaitRemaining(long timeoutMillis) throws InterruptedException {
+        awaitRemaining(Duration.ofMillis(timeoutMillis));
+    }
+
+    /**
+     * Disables traker and waits for tracked futures, cancels futures after timeout.
+     * Cancels all futures without wait if passed 0.
+     * Waits without timeout if passed negative.
+     *
+     * @param timeout duration to wait
      */
-    public void awaitRemaining (long timeoutMillis) {
-        long timeoutNanos = timeoutMillis * 1_000_000;
+    public void awaitRemaining(Duration timeout) throws InterruptedException {
+        long timeoutNanos = timeout.toNanos();
         long startNanos = System.nanoTime();
-        this.enabled = false;
 
         List<CompletableFuture<T>> remainingFutures;
-        synchronized (futures) {
-            if (futures.isEmpty())
+        lock.writeLock().lock();
+        try {
+            this.enabled = false;
```

Comment:
question: `enabled` is a volatile field. Why do we need to update it under the lock? If we want to make sure that we don't add the future to the tracking list if collection is disabled we should check if `enabled == true` inside `track` methods under the lock